presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java

```diff
@@ -121,16 +164,41 @@ public void testSubstring()
 
         assertFunction("SUBSTRING('Quadratically' FROM 5 FOR 6)", VARCHAR, "ratica");
         assertFunction("SUBSTRING('Quadratically' FROM 5 FOR 50)", VARCHAR, "ratically");
+        //
+        // Test SUBSTRING for non-ASCII
+        assertFunction("SUBSTRING('\u4FE1\u5FF5,\u7231,\u5E0C\u671B' FROM 1 FOR 1)", VARCHAR, "\u4FE1");
+        assertFunction("SUBSTRING('\u4FE1\u5FF5,\u7231,\u5E0C\u671B' FROM 3 FOR 5)", VARCHAR, ",\u7231,\u5E0C\u671B");
+        assertFunction("SUBSTRING('\u4FE1\u5FF5,\u7231,\u5E0C\u671B' FROM 4)", VARCHAR, "\u7231,\u5E0C\u671B");
+        assertFunction("SUBSTRING('\u4FE1\u5FF5,\u7231,\u5E0C\u671B' FROM -2)", VARCHAR, "\u5E0C\u671B");
+        assertFunction("SUBSTRING('\uD801\uDC2Dend' FROM 1 FOR 1)", VARCHAR, "\uD801\uDC2D");
+        assertFunction("SUBSTRING('\uD801\uDC2Dend' FROM 2 FOR 3)", VARCHAR, "end");
     }
 
     @Test
     public void testSplit()
     {
         assertFunction("SPLIT('a.b.c', '.')", new ArrayType(VARCHAR), ImmutableList.of("a", "b", "c"));
+
+        assertFunction("SPLIT('ab', '.', 1)", new ArrayType(VARCHAR), ImmutableList.of("ab"));
+        assertFunction("SPLIT('a.b', '.', 1)", new ArrayType(VARCHAR), ImmutableList.of("a.b"));
+        assertFunction("SPLIT('a.b.c', '.')", new ArrayType(VARCHAR), ImmutableList.of("a", "b", "c"));
         assertFunction("SPLIT('a..b..c', '..')", new ArrayType(VARCHAR), ImmutableList.of("a", "b", "c"));
         assertFunction("SPLIT('a.b.c', '.', 2)", new ArrayType(VARCHAR), ImmutableList.of("a", "b.c"));
         assertFunction("SPLIT('a.b.c', '.', 3)", new ArrayType(VARCHAR), ImmutableList.of("a", "b", "c"));
         assertFunction("SPLIT('a.b.c', '.', 4)", new ArrayType(VARCHAR), ImmutableList.of("a", "b", "c"));
+        assertFunction("SPLIT('a.b.c.', '.', 4)", new ArrayType(VARCHAR), ImmutableList.of("a", "b", "c"));
+        assertFunction("SPLIT('a.b.c.', '.', 3)", new ArrayType(VARCHAR), ImmutableList.of("a", "b", "c."));
+        // Test SPLIT for non-ASCII
+        assertFunction("SPLIT('\u4FE1\u5FF5,\u7231,\u5E0C\u671B', ',', 3)", new ArrayType(VARCHAR), ImmutableList.of("\u4FE1\u5FF5", "\u7231", "\u5E0C\u671B"));
+        assertFunction("SPLIT('\u8B49\u8BC1\u8A3C', '\u8BC1', 2)", new ArrayType(VARCHAR), ImmutableList.of("\u8B49", "\u8A3C"));
+
+        assertFunction("SPLIT('.a.b.c', '.', 4)", new ArrayType(VARCHAR), ImmutableList.of("", "a", "b", "c"));
+        assertFunction("SPLIT('.a.b.c', '.', 3)", new ArrayType(VARCHAR), ImmutableList.of("", "a", "b.c"));
+        assertFunction("SPLIT('.a.b.c', '.', 2)", new ArrayType(VARCHAR), ImmutableList.of("", "a.b.c"));
+        assertFunction("SPLIT('a..b..c', '.', 3)", new ArrayType(VARCHAR), ImmutableList.of("a", "", "b..c"));
+        assertFunction("SPLIT('a.b..', '.', 3)", new ArrayType(VARCHAR), ImmutableList.of("a", "b", "."));
```

Comment:
After discussion with @martint, the behavior of regexp split was changed to allow trailing empty part. I'm not sure if it's a good idea to keep the behavior of this split and split of regexp the same.